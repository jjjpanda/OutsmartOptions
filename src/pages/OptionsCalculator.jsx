import React from 'react';
import {
  Input,
  InputNumber,
  Button,
  Switch,
  Modal,
  Table,
  Collapse,
  Checkbox,
  Icon,
  Menu,
  Popover,
  Calendar,
  Spin,
} from 'antd';

import Tour from 'reactour';
import Cookie from 'js-cookie';
import { disableBodyScroll, enableBodyScroll } from 'body-scroll-lock';
import {
  aesthetique as aes,
  mathematique as math,
  utilique as util,
} from 'que-series';
import * as moment from 'moment';
import { NoAxisGraph, ProfitGraph } from '../components/Graphs.jsx';
import { StrategyInfo } from '../components/StrategyInfo.jsx';
import OptionsChain from '../components/OptionsChain.jsx'
import IVSkew from '../components/IVSkew.jsx'
import EquityModal from '../components/EquityModal.jsx'
import CalculateMenu from '../components/CalculateMenu.jsx'
import ReportModal from '../components/ReportModal.jsx'
import StockSymbol from '../components/StockSymbol.jsx';
import StockCalendar from '../components/StockCalendar.jsx';
import StrategySelector from '../components/StrategySelector.jsx'
import OptionsLeg from '../components/OptionsLeg.jsx';
import verifyUser from '../components/UserVerifier.jsx';

// JS Libraries

import {tutorialSteps} from './tour/OptionCalculatorTour.jsx'

const optionsMath = math.options;
const { treasury } = math;

const percentageColor = aes.color;

const { request, structure } = util;

const ButtonGroup = Button.Group;
const CollapsePanel = Collapse.Panel;

// Treasury Yields
let yields = [];
request.postFetchReq('/api/market/yields', '', (data) => {
  yields = data.yields;
});

class OptionsCalculator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      symbol: '',
      exists: true,
      priceChange: 0,
      price: 0,
      divYield: 0,
      addLegModalVisible: false,
      ivSkewModalVisible: false,
      isTourOpen: false,
      optionsChain: [['Empty', {}]],
      optionsSelected: [],
      numberIntervals: 15,
      percentInterval: 1,
      erVisible: false,
      calculateMenuVisible: false,
      reportLoading: false,
      editLegLoading: [],
      saved: false
    };
    verifyUser(({ loggedIn, user, email }) => {
      this.setState(() => ({ loggedIn }));
    });
    this.stockSymbol = React.createRef()
    
    this.props.updateApp(this.state)
  }

  updateSearchResults = (state, callback) => {
    console.log(state)
    this.setState(() => ({
      symbol: state.symbol,
      exists: state.exists,
      priceChange: state.priceChange,
      price: state.price,
      optionsChain: state.optionsChain,
      divYield: state.divYield,
      optionsSelected: [],
      earningsDate: state.earningsDate,
    }), () => {
      if(callback != undefined) {
        callback()
      }
    });
  }

  handleChange = (e) => {
    this.setState(() => ({ [e.target.id]: e.target.value }));
    console.log(this.state);
  }

  renderLegs() {
    return this.state.optionsSelected.map((option, index) => (
      <OptionsLeg isFirst={index === 0} key={option.key} callback={this.optionsSelectedMoreInfo} deleteSelf={this.deleteOption} optionRepresented={option} />
    ));
  }

  optionsSelectedMoreInfo = (e) => {
    console.log(e);
    this.setState((state) => ({ optionsSelected: [...state.optionsSelected.filter((item) => !(item.key == e.key)), e] }), () => this.resortOptionsSelected(e.symbol));
  }

  modalTrackSelected = (date) => {
    if (this.state.optionsSelected.map((e) => e.date).includes(date)) {
      return (
        <Icon type="pushpin" theme="twoTone" />
      );
    }

    return null;
  }


  resortOptionsSelected = (symbol) => {
    this.setState((state) => ({
      optionsSelected: [...state.optionsSelected].sort((a, b) => {
        const t = moment(a.date).diff(moment(b.date), 'hours');
        if (t > 0) {
          return 1;
        }
        if (t < 0) {
          return -1;
        }
        if (a.strike < b.strike) {
          return 1;
        }
        if (b.strike < a.strike) {
          return -1;
        }
        if (a.isCall) {
          return 1;
        }

        return -1;
      }),
    }), () => {
      setTimeout(() => {
        this.setState((state) => ({ editLegLoading: state.editLegLoading.filter((leg) => leg != (symbol)) }));
      }, 1000);
    });
  }

  addOption = (isCall, isLong, strike, price, cost, date, iv, symbol) => {
    this.setState((state) => ({
      optionsSelected: [...state.optionsSelected, {
        key: symbol, isCall, isLong, date, strike, price, cost, iv, symbol
      }],
    }), () => this.resortOptionsSelected(symbol));
  }

  deleteOption = (symbol) => {
    this.setState((state) => ({ optionsSelected: state.optionsSelected.filter((e) => !(e.key == symbol)) }), () => this.resortOptionsSelected(symbol));
  }

  loadInOptionsSelected = (legs) => {
    for(let leg of legs){
      const rfir = treasury.getRightYield(yields || [], moment(leg.date).diff(moment(), 'days')) / 100;
      const iv = optionsMath.calculateIV(moment(leg.date).diff(moment(), 'hours')/(365*24), leg.price, this.state.price, leg.strike, leg.isCall, rfir, this.state.divYield)
      const currentPrice = this.state.optionsChain.find(o => o[0] === leg.date)[1].find(o => o.strike === leg.strike)[`${leg.isCall ? "call" : "put"}`]
      console.log(currentPrice, leg.price)
      this.onHandleOptionLegChange(true, leg.isCall, leg.isLong, leg.strike, currentPrice, leg.price, leg.date, iv, leg.symbol)
    }

  }

  onHandleOptionLegChange = (needToAdd, isCall, isLong, strike, price, cost, date, iv, symbol) => {
    // console.log(needToAdd)
    // console.log((needToAdd ? "ADDING" : "DELETING")+' '+(isCall ? "Call" : "Put") + ' STRIKE: ' + strike + '@'+ price + ' => ' + date)
    if (needToAdd) {
      this.addOption(isCall, isLong, strike, price, cost, date, iv, symbol);
      this.setState((state) => ({ editLegLoading: [...state.editLegLoading, symbol] }), () => console.log(this.state.editLegLoading));
    } else {
      this.deleteOption(symbol);
      this.setState((state) => ({ editLegLoading: [...state.editLegLoading, symbol] }), () => console.log(this.state.editLegLoading));
    }
  }

  profitTableFormatting = () => {
    this.setState((state) => ({
      profitTableData: structure.dataToTableConversion(state.mergedOptions.profit),
    }));
    this.setState((state) => ({
      profitColumns: this.columnCreation(state.mergedOptions.profit, state.mergedOptions.percentProfit),
    }));
  }

  profitGraphFormatting = () => {
    this.setState((state) => ({
      profitGraphData: structure.dataToGraphConversion(state.optionsSelected.map((option) => ({ profit: option.profit, key: option.key }))),
    }),
    () => console.log(this.state));
  }

  columnCreation = (data, pData) => {
    const columns = [{ title: 'Price', dataIndex: 'x', fixed: 'left' }, ...data.map((key, i) => ({
      title: key[0],
      dataIndex: key[0],
      render: (text, record, index) =>
        //console.log(data[i][1][index][1], pData[i][1][index][1])
        (
          <div style={{ backgroundColor: percentageColor.hexColorFromPercentA([...pData[i][1]].reverse()[index][1]) }}>
            {text}
          </div>
        ),
    })),
    ];
    return columns;
  }

  calculateProfits = () => {
    const selectedOptions = this.state.optionsSelected;
    if (selectedOptions.filter((o) => !o.hide).length <= 0) {
      return;
    }
    
    for (const option of selectedOptions) {
      const rfir = treasury.getRightYield(yields || [], moment(option.date).diff(moment(), 'days')) / 100;
      option.greeks = optionsMath.calculateGreeks(moment(option.date).diff(moment(), 'hours') / (365*24), this.state.price, option.strike, option.isCall, option.isLong, rfir, this.state.divYield, option.iv);
      option.profit = optionsMath.calculateProfits(this.state.price, this.state.percentInterval, this.state.numberIntervals, option, rfir, this.state.divYield)
    }
    this.setState(() => ({ optionsSelected: selectedOptions }),
      () => {
        this.mergeOptions(selectedOptions);
      });
  }

  mergeOptions = (selectedOptions) => {
    const mergedOptions = {
      cost: 0,
      date: '',
      greeks: {
        delta: 0, gamma: 0, theta: 0, vega: 0, rho: 0,
      },
      profit: [],
    };

    for (const option of selectedOptions) {
      mergedOptions.cost += (option.isLong ? 1 : -1) * parseFloat(option.cost) * (option.hide ? 0 : parseInt(option.quantity));

      mergedOptions.greeks.delta += option.greeks.delta * option.hide ? 0 : option.quantity;
      mergedOptions.greeks.gamma += option.greeks.gamma * option.hide ? 0 : option.quantity;
      mergedOptions.greeks.theta += option.greeks.theta * option.hide ? 0 : option.quantity;
      mergedOptions.greeks.vega += option.greeks.vega * option.hide ? 0 : option.quantity;
      mergedOptions.greeks.rho += option.greeks.rho * option.hide ? 0 : option.quantity;
    }

    const optionsProfits = selectedOptions.filter((o) => !o.hide).map((o) => o.profit);

    mergedOptions.date = moment(selectedOptions.filter((o) => !o.hide).map((o) => moment(o.date).format('YYYY-MM-DD')).sort((a, b) => moment(a).diff(moment(b)))[0]).format('YYYY-MM-DD');

    mergedOptions.profit = optionsMath.mergeProfits(this.state.price, this.state.percentInterval, this.state.numberIntervals, optionsProfits, mergedOptions.date);
    
    mergedOptions.percentProfit = optionsMath.percentProfit(mergedOptions.profit, mergedOptions.cost)

    this.profitGraphFormatting();

    this.setState(() => ({ mergedOptions }),
      () => {
        this.profitTableFormatting();
        console.log(this.state);
      });
  }

  

  startTutorial = () => {
    // introJs('.intro').start();
    this.setState(() => ({ isTourOpen: true }));
  }

  closeTutorial = () => {
    this.setState(() => ({ isTourOpen: false }));
  }

  render() {
    return (
      <div>
        <div style={{ width: '60px', paddingBottom: '20px' }} />
        <div style={{ width: '60px', display: 'inline-block' }} />
        <h1 key="mainTitle" step-name="title" style={{ width: '135px', display: 'inline-block' }}>Outsmart Options</h1>
        <StockSymbol ref={this.stockSymbol} updateCallback={this.updateSearchResults} yieldCurve={yields} options historical={false} />

        <hr id="hr" align="left" />

        <div className="optionsList" step-name="example-contract">{this.renderLegs()}</div>

        <div className="optionsButtons">
          <div style={{ width: '15px', display: 'inline-block' }} />
          {/* <EquityModal symbol={this.state.symbol} price={this.state.price} onHandleOptionLegChange={this.onHandleOptionLegChange} />
           */}
          <div style={{ width: '43px', display: 'inline-block' }} />
          <OptionsChain onHandleOptionLegChange={this.onHandleOptionLegChange} modalTrackSelected={this.modalTrackSelected} editLegLoading={this.state.editLegLoading} optionsSelected={this.state.optionsSelected} optionsChain= {this.state.optionsChain} />
          <div style={{ width: '43px', display: 'inline-block' }} />
          <IVSkew modalTrackSelected={this.modalTrackSelected} optionsChain={this.state.optionsChain} />
          <div style={{ width: '43px', display: 'inline-block' }} />
          <div id="strategyButton"><Button icon="fund" onClick={() => {}}>Strategy</Button></div>
          <div style={{ width: '43px', display: 'inline-block' }} />
          <div id="calculateButton" step-name="calculate-button">
            <ButtonGroup>
              <Button onClick={this.calculateProfits} type="primary">Calculate</Button>
              <Button type="primary" icon="setting" onClick={() => { this.setState(() => ({ calculateMenuVisible: true })); }} />
            </ButtonGroup>
          </div>
          <Modal
            title="Calculate Settings"
            visible={this.state.calculateMenuVisible}
            onOk={() => { this.setState(() => ({ calculateMenuVisible: false })); }}
            onCancel={() => { this.setState(() => ({ calculateMenuVisible: false })); }}
          >
            <CalculateMenu 
              calculateProfits={this.calculateProfits} 
              intervalChange={(e) => { this.setState(() => ({ percentInterval: e })); }} 
              numberChange ={(e) => { this.setState(() => ({ numberIntervals: e })); }} 
            />
          </Modal>
          <div id="strategyButtons"><StrategySelector loadInOptionsSelected={this.loadInOptionsSelected} forceSearch={(symbol, callback) => {this.stockSymbol.current.onSearch(symbol, undefined, callback)}} symbol={this.state.symbol} optionsSelected={this.state.optionsSelected} /></div>
          <div id="calendarButton"><StockCalendar earningsDate={this.state.earningsDate} /></div>
        </div>
        <br />
        <div>
          {
          this.state.mergedOptions != undefined
            ? (
              <div>
                <div className="costStrategy">
                  <StrategyInfo stockPrice={this.state.price} optionsSelected={this.state.optionsSelected} mergedOptions={this.state.mergedOptions} />
                </div>

                <div className="profitGraphWrapper" step-name="profit-graph">
                  <ProfitGraph data={this.state.profitGraphData} keys={Object.keys(this.state.profitGraphData[0]).filter((o) => o != 'x')} />
                </div>

                <hr id="hr2" />
                <h3 style={{ marginLeft: '60px' }}>Profit Table:</h3>
                <div className="profitTableWrapper" step-name="profit-table" style={{ width: '80vw' }}>
                  <Table dataSource={this.state.profitTableData} columns={this.state.profitColumns} pagination={false} scroll={{ x: 'max-content' }} size="small" />
                </div>
                <ReportModal optionsSelected={this.state.optionsSelected} loading ={this.state.reportLoading} />
              </div>
            )
            : null
        }
        </div>
        <Tour
          steps={tutorialSteps(this.state, this)}
          showNavigation={false}
          showNumber={false}
          showButtons={false}
          showCloseButton={false}
          disableKeyboardNavigation
          maskSpace={3}
          startAt={0}
          update={JSON.stringify(this.state)}
          onAfterOpen={(target) => disableBodyScroll(target)}
          onBeforeClose={(target) => enableBodyScroll(target)}
          isOpen={this.state.isTourOpen}
          onRequestClose={this.closeTutorial}
        />
      </div>
    );
  }
}

export default OptionsCalculator;
