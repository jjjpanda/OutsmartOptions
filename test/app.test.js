import 'isomorphic-fetch'
import React from 'react'
import OptionsCalculator from "../src/calc.js"
import renderer from 'react-test-renderer'

describe('Jest Test', () => {
    it('sums numbers', () => {
        expect(1 + 2).toEqual(3);
        expect(2 + 2).toEqual(4);
     });
     test('Calculator', () => {
        const component = renderer.create(
          <OptionsCalculator />,
        );
        let tree = component.toJSON();
        expect(tree).not.toBe("");
      });
});

