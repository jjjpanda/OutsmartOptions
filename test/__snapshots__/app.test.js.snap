// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Test Calculator 1`] = `
<div>
  <h1
    style={
      Object {
        "paddingLeft": "60px",
        "paddingTop": "20px",
      }
    }
  >
    Outsmart Options
  </h1>
  <div>
    <div
      className="stockSymbol"
    >
      <div
        id="stockSymbolHeading"
      >
        Stock Symbol:
      </div>
      <div
        id="stockSymbolInput"
      >
        <div
          id="searchWrapper"
        >
          <span
            className="ant-input-search ant-input-affix-wrapper"
          >
            <input
              className="ant-input"
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Enter..."
              style={null}
              type="text"
              value=""
            />
            <span
              className="ant-input-suffix"
            >
              <i
                aria-label="icon: search"
                className="anticon anticon-search ant-input-search-icon"
                onClick={[Function]}
                tabIndex={-1}
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="search"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0 0 11.6 0l43.6-43.5a8.2 8.2 0 0 0 0-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                  />
                </svg>
              </i>
            </span>
          </span>
        </div>
        <div
          id="exists"
        />
      </div>
    </div>
    <div
      className="stockPrice"
    >
      <div
        id="stockPriceHeading"
      >
        Stock Price:
      </div>
      <div
        id="stockPriceBox"
      >
        <input
          className="ant-input ant-input-disabled"
          disabled={true}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="$0"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="stockPriceChange"
    >
      <div
        id="priceChangeHeading"
      >
        Stock Price Change:
      </div>
      <div
        id="priceChangeBox"
      >
        <input
          className="ant-input ant-input-disabled"
          disabled={true}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="0%"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <hr
    align="left"
    id="hr"
  />
  <div
    className="optionsList"
  />
  <div
    className="optionsButtons"
  >
    <div
      id="addLegButton"
    >
      <button
        className="ant-btn"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        <i
          aria-label="icon: edit"
          className="anticon anticon-edit"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="edit"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 0 0 0-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 0 0 9.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
            />
          </svg>
        </i>
        <span>
          Edit Legs
        </span>
      </button>
      <div
        className="addLegButtonWrapper"
      />
    </div>
    <div
      id="ivSkewButton"
    >
      <button
        className="ant-btn"
        onClick={[Function]}
        type="button"
      >
        <i
          aria-label="icon: profile"
          className="anticon anticon-profile"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="profile"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656zM492 400h184c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H492c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8zm0 144h184c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H492c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8zm0 144h184c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H492c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8zM340 368a40 40 0 1 0 80 0 40 40 0 1 0-80 0zm0 144a40 40 0 1 0 80 0 40 40 0 1 0-80 0zm0 144a40 40 0 1 0 80 0 40 40 0 1 0-80 0z"
            />
          </svg>
        </i>
        <span>
          IV Skew
        </span>
      </button>
    </div>
    <div
      id="strategyButton"
    >
      <button
        className="ant-btn"
        onClick={[Function]}
        type="button"
      >
        <i
          aria-label="icon: fund"
          className="anticon anticon-fund"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="fund"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M926 164H94c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V196c0-17.7-14.3-32-32-32zm-40 632H134V236h752v560zm-658.9-82.3c3.1 3.1 8.2 3.1 11.3 0l172.5-172.5 114.4 114.5c3.1 3.1 8.2 3.1 11.3 0l297-297.2c3.1-3.1 3.1-8.2 0-11.3l-36.8-36.8a8.03 8.03 0 0 0-11.3 0L531 565 416.6 450.5a8.03 8.03 0 0 0-11.3 0l-214.9 215a8.03 8.03 0 0 0 0 11.3l36.7 36.9z"
            />
          </svg>
        </i>
        <span>
          Strategy
        </span>
      </button>
    </div>
    <div
      id="calculateButton"
    >
      <button
        className="ant-btn ant-btn-primary"
        onClick={[Function]}
        type="button"
      >
        <span>
          Calculate
        </span>
      </button>
    </div>
    <div
      id="saveButton"
    >
      <button
        className="ant-btn ant-btn-circle ant-btn-icon-only"
        onClick={[Function]}
        type="button"
      >
        <i
          aria-label="icon: save"
          className="anticon anticon-save"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="save"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M893.3 293.3L730.7 130.7c-7.5-7.5-16.7-13-26.7-16V112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V338.5c0-17-6.7-33.2-18.7-45.2zM384 184h256v104H384V184zm456 656H184V184h136v136c0 17.7 14.3 32 32 32h320c17.7 0 32-14.3 32-32V205.8l136 136V840zM512 442c-79.5 0-144 64.5-144 144s64.5 144 144 144 144-64.5 144-144-64.5-144-144-144zm0 224c-44.2 0-80-35.8-80-80s35.8-80 80-80 80 35.8 80 80-35.8 80-80 80z"
            />
          </svg>
        </i>
      </button>
    </div>
    <div
      id="savedStrategyButton"
    >
      <button
        className="ant-btn ant-btn-circle ant-btn-icon-only"
        onClick={[Function]}
        type="button"
      >
        <i
          aria-label="icon: download"
          className="anticon anticon-download"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="download"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M505.7 661a8 8 0 0 0 12.6 0l112-141.7c4.1-5.2.4-12.9-6.3-12.9h-74.1V168c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v338.3H400c-6.7 0-10.4 7.7-6.3 12.9l112 141.8zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
            />
          </svg>
        </i>
      </button>
    </div>
  </div>
  <br />
  <div>
    <pre />
  </div>
</div>
`;
